org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/validation/XbaseValidator.java

```diff
@@ -401,24 +401,29 @@ public void checkTypeReferenceIsNotVoid(XCasePart expression) {
 	
 	@Check
 	public void checkReturn(XReturnExpression expr) {
+		XExpression returnedExpression = expr.getExpression();
```

Comment:
My gut feeling is that this is more a type system issue and should be handled there. When the type for a return expression is computed, the state likely contains information that we are in the context of another return expression.